### PR TITLE
Add `emulation` test plan

### PR DESCRIPTION
**Pull request by ackintosh**
## How to run

```shell
# Import test plans
git clone https://github.com/sigp/gossipsub-testground.git
testground plan import --from ./gossipsub-testground/

# Run `emulation` test plan
cd gossipsub-testground
testground run composition -f emulation/compositions/emulation.toml --wait
```

#### Test parameters available

- Please see [manifest.toml](https://github.com/ackintosh/gossipsub-testground/blob/test-plan-emulation/emulation/manifest.toml#L17-L21)

#### READMEs

In this PR, the following READMEs have been updated/added.

- [README](https://github.com/ackintosh/gossipsub-testground/blob/test-plan-emulation/README.md#dashboards)
- [emulation/README](https://github.com/ackintosh/gossipsub-testground/blob/test-plan-emulation/emulation/README.md)

## What the emulation does

Note: Attackers connect to a single publisher (victim).

image

## Dashboards

Please run `docker-compose up` and browse http://localhost:13000/dashboards.

### Gossipsub Metrics

The metrics of gossipsub are recorded once the emulation has been completed. [All of the metrics on libp2p-gossipsub](https://github.com/ackintosh/gossipsub-testground/blob/test-plan-emulation/emulation/src/honest.rs#L235-L275) are available in this dashboard.

Variables for this dashboard:

- `run_id`: The run_id for the test run you want to see.
- `topic`: The gossipsub topic, currently it's fixed to `emulate`.
- `instance_name`: Some panels in this dashboard need an instance name to be specified. (e.g. score_per_mesh histogram)

image

### Peer Scores

The peer scores are recorded periodically (every second) while the emulation is running. 

Variables for this dashboard:

- `run_id`: The run_id for the test run you want to see.
- `instance_name`: It's default to `All`, you can select specific instances.

image

---

## Remaining work

- ~~`prometheus-client`: Finish~~ https://github.com/prometheus/client_rust/pull/83. -> Done
- `libp2p-gossipsub`: Make a change to enable the protobuf support.
  - Note: In this PR, the forked ones are used for now. ([prometheus-client](https://github.com/ackintosh/gossipsub-testground/blob/test-plan-emulation/emulation/Cargo.toml#L30-L32), [libp2p](https://github.com/ackintosh/gossipsub-testground/blob/test-plan-emulation/emulation/Cargo.toml#L12-L14))
  - Please see [here](https://github.com/sigp/gossipsub-testground/pull/4#issuecomment-1250415907) for the current progress.
- ~~Create Grafana dashboard~~ -> Done

### CODE REVIEW

**Comment by ackintosh** [comment]
Regarding recording gossipsub metrics: `prometheus-client` is used in libp2p-gossipsub, but in testground, InfluxDB. So in order to record gossipsub metrics (stored in `prometheus-client`) to InfluxDB, I need to progress [Protobuf support](https://github.com/prometheus/client_rust/pull/47) on `prometheus-client`.

**Comment by ackintosh** [comment]
There are still remaining works (see [`Remaining work`](https://github.com/sigp/gossipsub-testground/pull/4#issue-1310275729)), now the `emulation` test plan can be run.

**Comment by divagant-martian** [comment]
do you have the source of the mermaid diagram? It would be good to include it as code in a README.md of the crate to have it beyond of as an image in the PR

**Review by divagant-martian — COMMENTED** [review]
Looking pretty good. Would you mind changing the comments of constants and functions to docs? `// Some docs` to `/// Some docs`

**Comment by ackintosh** [comment]
> do you have the source of the mermaid diagram? It would be good to include it as code in a README.md of the crate to have it beyond of as an image in the PR

The diagram is created using drawio so I'll try to write the diagram using mermaid.

File: Cargo.toml [hunk 1]
```diff
@@ -8,4 +8,5 @@ members = [
 # should have its Cargo.lock.
 exclude = [
     "smoke",
+    "emulation",
```
[the remark this answers is not in the document]
Comment (ackintosh) [reply Cargo.toml:11]: Oh, you are right. 💡

File: emulation/src/utils.rs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+    for metric in family.metrics.iter() {
+        let histogram = get_histogram_value(metric);
+        for bucket in histogram.buckets.iter() {
+            let mut query = WriteQuery::new(Local::now().into(), family.name.clone())
+                .add_tag("instance_peer_id", instance_info.peer_id.to_string())
+                .add_tag("instance_name", instance_info.name())
+                .add_tag("run_id", run_id.to_owned())
+                .add_field("count", bucket.count)
+                .add_field("upper_bound", bucket.upper_bound);
```
[the remark this answers is not in the document]
Comment (ackintosh) [reply emulation/src/utils.rs:158]: Ahhh, sorry I'm not sure if it's possible. I put off that as I'm not familiar with Grafana...

[the remark this answers is not in the document]
Comment (ackintosh) [reply emulation/src/utils.rs:158]: I will learn Grafana and may change this. Please let me know if you have concerns. 🙏

File: emulation/src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        )
+        .map(|event| debug!("Event: {:?}", event))
+        .collect::<Vec<()>>()
+        .await;
+
+    Ok(())
+}
+
+// Create InfluxDB queries for Counter metrics.
+pub(crate) fn queries_for_counter(
```
[the remark this answers is not in the document]
Comment (ackintosh) [reply emulation/src/utils.rs:86]: Do you mean [these tags (instance_peer_id, instance_name)](https://github.com/ackintosh/gossipsub-testground/blob/eeb25cee9af8913e39f80f2fd603f1e89b98e294/emulation/src/utils.rs#L95-L96) are not necessary?

[the remark this answers is not in the document]
Comment (divagant-martian) [reply emulation/src/utils.rs:86]: Question just out of curiosity. Code looks good as is

**Comment by ackintosh** [comment]
### Trying to enable the `protobuf` feature on libp2p

I have updated libp2p on https://github.com/sigp/gossipsub-testground/pull/4/commits/e46643a094d600ddc24f520c876835a1510474bc. The revision of libp2p `7eb216f4c259c8f5b6f23dbdc044b45e8a38c2e5` is in https://github.com/libp2p/rust-libp2p/pull/2911.

---

The compile error occurs now:

```
error[E0308]: mismatched types
   --> emulation/src/honest.rs:287:23
    |
287 |                 Some((registry, Config::default())),
    |                       ^^^^^^^^ expected trait `SendSyncEncodeMetric`, found trait `prometheus_client::encoding::proto::EncodeMetric`
    |
    = note: expected mutable reference `&mut prometheus_client::registry::Registry`
               found mutable reference `&mut prometheus_client::registry::Registry<Box<(dyn prometheus_client::encoding::proto::EncodeMetric + 'static)>>`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `emulation` due to previous error
```

I think something updates are needed on `libp2p` side, but I'm stuck. 🤔

File: emulation/src/utils.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ackintosh) [reply emulation/src/utils.rs:158]: I noticed that [Flux](https://docs.influxdata.com/influxdb/v1.8/flux/) should be enabled to create histograms with InfluxDB.

I've filed the PR to testground to enable Flux. https://github.com/testground/testground/pull/1473

[the remark this answers is not in the document]
Comment (ackintosh) [reply emulation/src/utils.rs:158]: Note for self: How to represent the buckets to histogram on Grafana (using Bar chart).

- Bar chart
- X Axis: upper_bound

## score_per_mesh

image


## heartbeat_duration

image

**Comment by ackintosh** [comment]
Note for self: TODO

invalid_messages_per_topic
accepted_messages_per_topic
ignored_messages_per_topic
rejected_messages_per_topic
topic_iwant_msgs

**Review by AgeManning — APPROVED** [review]
This is solid progress. Lets merge down and make adjustments in future PRs

**Pull request merged**